radix-engine-interface/src/time/date_time.rs

```diff
@@ -0,0 +1,512 @@
+                "Invalid minute. Expected a value between 0 (inclusive) and 59 (inclusive)"
+                    .to_string(),
+            );
+        }
+
+        if second > 59 {
+            return Err(
+                "Invalid second. Expected a value between 0 (inclusive) and 59 (inclusive)"
+                    .to_string(),
+            );
+        }
+
+        Ok(Self {
+            year,
+            month,
+            day_of_month,
+            hour,
+            minute,
+            second,
+        })
+    }
+
+    pub fn from_instant(instant: &Instant) -> Result<Self, String> {
+        if instant.seconds_since_unix_epoch < MIN_SUPPORTED_TIMESTAMP
+            || instant.seconds_since_unix_epoch > MAX_SUPPORTED_TIMESTAMP
+        {
+            return Err(format!(
+                "Instant out of supported range [{}, {}]",
+                MIN_SUPPORTED_TIMESTAMP, MAX_SUPPORTED_TIMESTAMP
+            ));
+        }
+
+        // First, convert the base to 1 Mar 2000 for easier leap year calculation
+        let secs_since_march_y2k: i64 =
+            instant.seconds_since_unix_epoch - SHIFT_FROM_UNIX_TIME_TO_MARCH_Y2K;
+
+        let mut days_since_march_y2k = secs_since_march_y2k / SECONDS_IN_A_DAY as i64;
+        let mut remaining_secs = secs_since_march_y2k % SECONDS_IN_A_DAY as i64;
+        if remaining_secs < 0 {
+            remaining_secs += SECONDS_IN_A_DAY as i64;
+            days_since_march_y2k -= 1;
+        }
+
+        let mut num_400_year_cycles = days_since_march_y2k / DAYS_PER_400Y;
+        let mut remaining_days = days_since_march_y2k % DAYS_PER_400Y;
+        if remaining_days < 0 {
+            remaining_days += DAYS_PER_400Y;
+            num_400_year_cycles -= 1;
+        }
+
+        let mut num_100_year_cycles = remaining_days / DAYS_PER_100Y;
+        if num_100_year_cycles == 4 {
+            // Subtract one due 400 years cycle (400 years cycle fits 4 100y cycles)
+            num_100_year_cycles -= 1;
+        }
+        remaining_days -= num_100_year_cycles * DAYS_PER_100Y;
+
+        let mut num_4_year_cycles = remaining_days / DAYS_PER_4Y;
+        if num_4_year_cycles == 25 {
+            // Subtract one due 100 years cycle (100 years cycle fits 25 4y cycles)
+            num_4_year_cycles -= 1;
+        }
+        remaining_days -= num_4_year_cycles * DAYS_PER_4Y;
+
+        let mut remaining_years = remaining_days / 365;
+        if remaining_years == 4 {
+            // Subtract one due to four years cycle
+            remaining_years -= 1;
+        }
+        remaining_days -= remaining_years * 365;
+
+        let mut year =
+            remaining_years +
+                4 * num_4_year_cycles +
+                100 * num_100_year_cycles +
+                400 * num_400_year_cycles
+                + 2000 /* Add the base year (after shifting) */;
+
+        let mut days_in_months_starting_on_march = LEAP_YEAR_DAYS_IN_MONTHS.clone();
+        days_in_months_starting_on_march.rotate_left(2);
```

Comment:
We don't consider non-leap year here?

Reply:
Yeah, this surely isn't obvious :sweat_smile: but `remaining_days` is at most 365 (29 Feb on leap years), and, thanks to shifting the base to March (see comment above `const SHIFT_FROM_UNIX_TIME_TO_MARCH_Y2K`) Feb comes last, so, after iterating through all other months, the while loop predicate for the last month [Feb]:  `days[..] <= remaining_days` never matches.

In other words, we don't care about the leap year for dates before Mar 1: if the month happens to be Feb we just end up with a correct number of remaining_days, and if the date is on or after Mar 1, then it's just another full year that's already been handled earlier.

Reply:
Woooow, that's very convoluted and interesting!